samples/nrf9160/nrf_cloud_rest_agps/src/main.c

```diff
@@ -0,0 +1,604 @@
+				snprintk(&print_buf[print_buf_len - 1],
+					 sizeof(print_buf) - print_buf_len,
+					 "%d  ", i + 1);
+			if (print_buf_len < 0) {
+				LOG_ERR("Failed to print satellite stats");
+				break;
+			}
+		}
+	}
+
+	LOG_INF("%s", log_strdup(print_buf));
+}
+
+static void gps_handler(const struct device *dev, struct gps_event *evt)
+{
+	ARG_UNUSED(dev);
+
+	switch (evt->type) {
+	case GPS_EVT_SEARCH_STARTED:
+		LOG_INF("GPS_EVT_SEARCH_STARTED");
+		break;
+	case GPS_EVT_SEARCH_STOPPED:
+		LOG_INF("GPS_EVT_SEARCH_STOPPED");
+		break;
+	case GPS_EVT_SEARCH_TIMEOUT:
+		LOG_INF("GPS_EVT_SEARCH_TIMEOUT");
+		break;
+	case GPS_EVT_OPERATION_BLOCKED:
+		LOG_INF("GPS_EVT_OPERATION_BLOCKED");
+		break;
+	case GPS_EVT_OPERATION_UNBLOCKED:
+		LOG_INF("GPS_EVT_OPERATION_UNBLOCKED");
+		break;
+	case GPS_EVT_AGPS_DATA_NEEDED:
+		LOG_INF("GPS_EVT_AGPS_DATA_NEEDED");
+		memcpy(&agps_request, &evt->agps_request, sizeof(agps_request));
+		k_sem_give(&do_agps_request_sem);
+		break;
+	case GPS_EVT_PVT:
+		print_satellite_stats(&evt->pvt);
+		break;
+	case GPS_EVT_PVT_FIX:
+		LOG_INF("GPS_EVT_PVT_FIX");
+		print_pvt_data(&evt->pvt);
+		break;
+	case GPS_EVT_NMEA_FIX:
+		LOG_INF("GPS_EVT_NMEA_FIX");
+		break;
+	default:
+		break;
+	}
+}
+
+int set_led(const int state)
+{
+#if defined(CONFIG_REST_ENABLE_LED)
+	int err = dk_set_led(LED_NUM, state);
+
+	if (err) {
+		LOG_ERR("Failed to set LED, error: %d", err);
+		return err;
+	}
+#else
+	ARG_UNUSED(state);
+#endif
+	return 0;
+}
+
+int init_led(void)
+{
+#if defined(CONFIG_REST_ENABLE_LED)
+	int err = dk_leds_init();
+
+	if (err) {
+		LOG_ERR("LED init failed, error: %d", err);
+		return err;
+	}
+	(void)set_led(0);
+#endif
+	return 0;
+}
+
+static void button_handler(uint32_t button_states, uint32_t has_changed)
+{
+	if (has_changed & button_states &
+	    BIT(BUTTON_EVENT_BTN_NUM - 1)) {
+
+		LOG_DBG("Button %d pressed", BUTTON_EVENT_BTN_NUM);
+		k_sem_give(&button_press_sem);
+	}
+}
+
+static int request_jitp(void)
+{
+	int ret = 0;
+
+#if defined(CONFIG_REST_DO_JITP)
+	jitp_requested = false;
+
+	(void)k_sem_take(&button_press_sem, K_NO_WAIT);
+
+	LOG_INF("Press button %d to request just-in-time provisioning", BUTTON_EVENT_BTN_NUM);
+	LOG_INF("Waiting %d seconds..", JITP_REQ_WAIT_SEC);
+
+	ret = k_sem_take(&button_press_sem, K_SECONDS(JITP_REQ_WAIT_SEC));
+	if (ret == -EAGAIN) {
+		LOG_INF("JITP will be skipped");
+		ret = 0;
+	} else if (ret) {
+		LOG_ERR("k_sem_take: err %d", ret);
+	} else {
+		jitp_requested = true;
+		LOG_INF("JITP will be performed after network connection is obtained");
+		ret = 0;
+	}
+#endif
+	return ret;
+}
+
+static int do_jitp(void)
+{
+	int ret = 0;
+
+#if defined(CONFIG_REST_DO_JITP)
+	if (!jitp_requested) {
+		return 0;
+	}
+
+	LOG_INF("Performing JITP..");
+	ret = nrf_cloud_rest_jitp(CONFIG_NRF_CLOUD_SEC_TAG);
+
+	if (ret == 0) {
+		LOG_INF("Waiting 30s for cloud provisioning to complete..");
+		k_sleep(K_SECONDS(30));
+		(void)k_sem_take(&button_press_sem, K_NO_WAIT);
```

Comment:
Why double `k_sem_take()` call? If you need to "reset" semaphore state, there's a dedicated `k_sem_reset()` function, IMO more intuitive for such use case.